src/murmur/Messages.cpp

```diff
@@ -155,6 +155,61 @@ bool isChannelEnterRestricted(Channel *c) {
 	return false;
 }
 
+/// Process the Capabilities message, used to negotiate voice protocol.
+/// See the doc in Mumble.proto for details about this message.
+///
+/// @param uSource ServerUser that receives this message.
+/// @param msg The message itself.
+void Server::msgCapabilities(ServerUser *uSource, MumbleProto::Capabilities &msg) {
+	MSG_SETUP(ServerUser::Connected);
+	int custom_index = 0;
+	for (int i = 0; i < msg.supported_protocols_size(); ++i){
+		auto protocol_type = msg.supported_protocols(i);
+		switch (protocol_type) {
+			case MumbleProto::Capabilities_VoiceProtocol_UDP_OCB2:
+				uSource->qlSupportedVoiceProtocols.append(VoiceProtocol(VoiceProtocolType::UDP_OCB2));
+				break;
+			case MumbleProto::Capabilities_VoiceProtocol_CUSTOM:
+				uSource->qlSupportedVoiceProtocols.append(VoiceProtocol(msg.custom_protocols(custom_index)));
+				++custom_index;
+				break;
+			default:
+				continue;
+		}
+	}
+	uSource->vptVoiceProtocolType = VoiceProtocolType::UNDEFINED;
+	for (int i = 0; i < Meta::mp.qlAllowedVoiceProtocolTypes.count()
+		&& uSource->vptVoiceProtocolType == VoiceProtocolType::UNDEFINED; ++i) {
+
+		for (int j = 0; j < uSource->qlSupportedVoiceProtocols.count(); ++j) {
+			if (uSource->qlSupportedVoiceProtocols[j].protocolType == Meta::mp.qlAllowedVoiceProtocolTypes[i]) {
+				uSource->vptVoiceProtocolType = Meta::mp.qlAllowedVoiceProtocolTypes[i];
+				break;
+			}
+		}
+	}
```

Comment:
With this implementation I think the preference of the client is neglected. I think the loops should be changed so the server should try to use the first supported protocol the client suggests, if it is supported. If not, use the next.

Reply:
I was thinking the opposite way. I tend to put more weight on the servers' choice. Can you elaborate more?

Reply:
I think the server shouldn't care as the server is not a human being (and also not directly controlled by one) whereas the client is. Thus a user might for some reason prefer to use one protocol over another (maybe because that person thinks that it provides better voice quality) and if we give this choice to them, they'll see it as an extra gimmick and be happy. Otherwise I think people might complain that "the server supports X, so why can't I use it?".

All the server should care about is what protocols/ciphers it definitely doesn't want to use (because they are insecure or whatever). Other than that I can't imagine a server having a preference for one protocol over another :thinking:

Reply:
Okay. I think you have made a very good point. The server should focus on banning things, and respect clients' preference.